ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs

```diff
@@ -0,0 +1,320 @@
+
+-- | Predicate on pairs of ledger states of subsequent eras that are converted to property tests.
+data TranslationPredicate src dest = TranslationPredicate {
+    tpTestName :: TestName
+  , tpRun      :: LedgerState src EmptyMK -> LedgerState dest DiffMK -> Bool
+}
+
+-- Intermediate data type
+-- REVIEW: Can we get rid of this?
+data PreProperty x y = PreProperty {
+    ppTestName :: TestName
+  , ppRun      :: TestSetup x y -> Bool
+}
+
+toTestTrees
+  :: (AllPairs TestSetup Arbitrary xs, AllPairs TestSetup Show xs)
+  => InPairs TranslationPredicate xs
+  -> InPairs (RequiringBoth WrapLedgerConfig TranslateLedgerState) xs
+  -> [TestTree]
+toTestTrees predicates translations = mkTestTrees $ zipInPairs predicates translations
+  where
+    fn
+      :: TranslationPredicate x y
+      -> RequiringBoth WrapLedgerConfig TranslateLedgerState x y
+      -> PreProperty x y
+    fn TranslationPredicate {..} translateLedgerState = PreProperty {
+      ppTestName = tpTestName
+    , ppRun = \TestSetup {..} ->
+        let translation = provideBoth
+                translateLedgerState
+                (WrapLedgerConfig tsSrcLedgerConfig)
+                (WrapLedgerConfig tsDestLedgerConfig)
+            destState = translateLedgerStateWith translation tsEpochNo tsSrcLedgerState
+        in tpRun tsSrcLedgerState destState
+    }
+
+    zipInPairs
+      :: InPairs TranslationPredicate xs
+      -> InPairs (RequiringBoth WrapLedgerConfig TranslateLedgerState) xs
+      -> InPairs PreProperty xs
+    zipInPairs PNil PNil                 = PNil
+    zipInPairs (PCons b bs) (PCons z zs) = PCons (fn b z) (zipInPairs bs zs)
+
+    mkTestTrees
+      :: (AllPairs TestSetup Arbitrary xs, AllPairs TestSetup Show xs)
+      => InPairs PreProperty xs -> [TestTree]
+    mkTestTrees PNil                        = []
+    mkTestTrees (PCons PreProperty {..} cs) =
+      testProperty ppTestName ppRun : mkTestTrees cs
+
+{-------------------------------------------------------------------------------
+    Utilities
+-------------------------------------------------------------------------------}
+
+extractUtxoDiff
+  :: LedgerState (ShelleyBlock proto era) DiffMK
+  -> Diff (TxIn (EraCrypto era)) (Core.TxOut era)
+extractUtxoDiff shelleyLedgerState =
+  let ApplyDiffMK tables = shelleyUTxOTable $ shelleyLedgerTables shelleyLedgerState
+  in tables
+
+{-------------------------------------------------------------------------------
+    TestSetup
+-------------------------------------------------------------------------------}
+
+data TestSetup src dest = TestSetup {
+    tsSrcLedgerConfig  :: LedgerConfig src
+  , tsDestLedgerConfig :: LedgerConfig dest
+  , tsSrcLedgerState   :: LedgerState src EmptyMK
+  , tsEpochNo          :: EpochNo
+}
+
+deriving instance ( Show (LedgerConfig src)
+                  , Show (LedgerConfig dest)
+                  , Show (LedgerState src EmptyMK)) => Show (TestSetup src dest)
+
+-- todo: Useful to merge some of these instances?
```

Comment:
I don't see how we could merge the Byron to Shelley instances. I'm not sure about the others. Do you know what prevents us from defining a:

```haskell 
instance Arbitrary (TestSetup (ShelleyBlock p era0) (ShelleyBlock p era1))
```